client/core/core.go

```diff
@@ -4563,16 +4565,64 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("match proof note unmarshal error: %w", err)
 	}
 
-	// Expire the epoch
-	dc.setEpoch(note.MarketID, note.Epoch+1)
+	// Expire previous epoch.
+	dc.tryUpdateEpoch(note.MarketID, note.Epoch+1)
 
 	book := dc.bookie(note.MarketID)
 	if book == nil {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	err = validateMatchCSum(note.Preimages, note.CSum, dc.trackedTrades())
```

Comment:
The use of the two helper methods is overkill. Just iterate here.

```
for _, trade := range dc.trackedTrades() {
    if trade.metaData.Status == order.OrderStatusEpoch && len(trade.csum) > 0 && !bytes.Equal(note.CSum, trade.csum)
        // Probably self-revoke the order so we don't acknowledge match requests. Also log the error.
    }
}
```

Reply:
What's the purpose for this check `len(trade.csum) > 0` ? 

My understanding is - if `len(trade.csum) == 0` this validation must fail as well.

Reply:
That would imply dexc has restarted and csum was not restored (a rare event as restarting with epoch orders is a super dumb idea), or the trade is from a different epoch and a preimage request has not been received for that trade.  Related to my other comments regarding orders in OrderStatusEpoch potentially spanning multiple epochs.